src/io/delimited.cr

```diff
@@ -0,0 +1,124 @@
+module IO
+  # An IO that wraps another IO, and only reads up to the beginning of a
+  # specified delimiter.
+  #
+  # This is useful for exposing part of an underlying stream to a client.
+  #
+  # ```
+  # io = MemoryIO.new "abc||123"
+  # delimited = IO::Delimited.new(io, read_delimiter: "||")
+  #
+  # delimited.gets_to_end # => "abc"
+  # delimited.gets_to_end # => nil
+  # io.gets_to_end        # => "123"
+  # ```
+  class Delimited
+    include IO
+
+    # If `sync_close` is true, closing this IO will close the underlying IO.
+    property? sync_close
+
+    getter read_delimiter
+    getter? closed : Bool
+
+    @delimiter_buffer : Bytes
+    @active_delimiter_buffer : Bytes?
```

Comment:
Oops, I seem to have missed your comment for about 7 hours, only caught it when i was looking at my email.

I actually assumed that a slice of 0 bytes was impossible, and that it would throw if i tried `Slice[0] + 1`, so this working is a nice surprise!

It does indeed seem to pass all the specs, and the [benchmark](https://gist.github.com/RX14/33759091eedfbef1d390474140fcd6e1)... holy moly, it's twice as fast! I didn't realise that nil-checking was so expensive, can you explain why that is?